src/persist/src/indexed/encoding.rs

```diff
@@ -230,7 +235,36 @@ impl BlobTraceUpdates {
                 // Recurse at most once, since this data is now structured.
                 self.get_or_make_structured::<K, V>(key_schema, val_schema)
             }
-            BlobTraceUpdates::Both(_, structured) => structured,
+            BlobTraceUpdates::Both(_, structured) => {
+                // If the types don't match, attempt to migrate the array to the new type.
+                // We expect this to succeed, since this should only be called with backwards-
+                // compatible schemas... but if it fails we only log, and let some higher-level
+                // code signal the error if it cares.
+                let migrate = |array: &mut ArrayRef, to_type: DataType| {
+                    let from_type = array.data_type().clone();
+                    if from_type != to_type {
+                        if let Some(migration) = backward_compatible(&from_type, &to_type) {
+                            *array = migration.migrate(Arc::clone(array));
+                        } else {
+                            error!(
+                                ?from_type,
+                                ?to_type,
+                                "failed to migrate array type; backwards-incompatible schema migration?"
+                            );
+                        }
+                    }
+                };
+                migrate(
+                    &mut structured.key,
+                    key_schema.data_type().expect("valid key schema"),
```

Comment:
Does this mean we use the schema of the WriteHandle that happens to be running this compaction?

Reply:
It definitely should be the schema of the data compaction produces as output. I think at the moment those things are the same?

Once we support multiple schemas, we'll need some other logic for selecting the "real_schemas" we use during compaction... but whatever that logic ends up being, those are the schemas we'd want to pass through here too, so this logic ~wouldn't need to change.

Reply:
Ah! So this is meant to unblock turning on the columnar->columnar compaction code, but not yet enough to turn on schema changes.

> This together is hypothetically enough to migrate the data we have in staging.

You did say ☝️ in the PR description, but I read that differently at first. Makes sense now, thanks!

Reply:
Ah yeah - migrate is a bit overloaded and I should have been more precise!